image/image.go

```diff
@@ -0,0 +1,386 @@
+		})
+	}
+	if err := seedYaml.Write(); err != nil {
+		return fmt.Errorf("cannot write seed.yaml: %s", err)
+	}
+
+	// now do the bootloader stuff
+	if err := partition.InstallBootConfig(opts.GadgetUnpackDir); err != nil {
+		return err
+	}
+
+	if err := setBootvars(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setBootvars() error {
+	// set the bootvars for kernel/os snaps so that the system
+	// actually boots and can do the `firstboot` import of the snaps.
+	//
+	// there is also no mounted os/kernel snap, all we have are the
+	// blobs
+
+	bootloader, err := partition.FindBootloader()
+	if err != nil {
+		return fmt.Errorf("can not set kernel/os bootvars: %s", err)
+	}
+
+	snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDir, "*.snap"))
+	if len(snaps) == 0 {
+		return fmt.Errorf("internal error: cannot find os/kernel snap")
+	}
+	for _, fullname := range snaps {
+		bootvar := ""
+
+		// detect type
+		snapFile, err := snap.Open(fullname)
+		if err != nil {
+			return fmt.Errorf("can not read %v", fullname)
+		}
+		// read .sideinfo
+		var si snap.SideInfo
+		siFn := fullname + ".sideinfo"
+		if osutil.FileExists(siFn) {
+			j, err := ioutil.ReadFile(siFn)
+			if err != nil {
+				return err
+			}
+			if err := json.Unmarshal(j, &si); err != nil {
+				return fmt.Errorf("cannot read metadata: %s %s\n", siFn, err)
+			}
+		}
+		info, err := snap.ReadInfoFromSnapFile(snapFile, &si)
+		if err != nil {
+			return fmt.Errorf("can not get info for %v", fullname)
+		}
+		// local install
+		if info.Revision.Unset() {
+			info.Revision = snap.R(-1)
+		}
+
+		switch info.Type {
+		case snap.TypeOS:
+			bootvar = "snap_core"
+		case snap.TypeKernel:
+			bootvar = "snap_kernel"
+			if err := extractKernelAssets(fullname, info); err != nil {
+				return err
+			}
+		}
+
+		name := filepath.Base(fullname)
+		if bootvar != "" {
+			if err := bootloader.SetBootVar(bootvar, name); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func runCommand(cmdStr ...string) error {
+	cmd := exec.Command(cmdStr[0], cmdStr[1:]...)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("cannot run %v: %s (%s)", cmdStr, err, output)
+	}
+	return nil
+}
+
+func extractKernelAssets(snapPath string, info *snap.Info) error {
+	// FIXME: hrm, hrm, we need to be root for this - alternatively
+	//        we could make boot.ExtractKernelAssets() work on
+	//        a plain .snap file again (i.e. revert bee59a2)
+	if err := os.MkdirAll(info.MountDir(), 0755); err != nil {
+		return err
+	}
+	defer os.Remove(filepath.Dir(info.MountDir()))
+	defer os.Remove(info.MountDir())
+
+	if err := runCommand("mount", snapPath, info.MountDir()); err != nil {
+		return err
+	}
+	defer runCommand("umount", info.MountDir())
+
+	pb := progress.NewTextProgress()
+	if err := boot.ExtractKernelAssets(info, pb); err != nil {
+		return err
+	}
+	return nil
+}
+
+func copyLocalSnapFile(snapName, targetDir string) (string, *snap.Info, error) {
```

Comment:
Would be good to name the result, as it's not clear without reading the code what that string is.